packages/arb-bridge-eth/contracts/GlobalInbox.sol

```diff
@@ -229,95 +228,38 @@ contract GlobalInbox is GlobalEthWallet, GlobalFTWallet, GlobalNFTWallet, IGloba
-        require(_values.length == messageCount, "wrong input length");
-        require(_messageLengths.length == messageCount, "wrong input length");
-
-        Inbox storage inbox = inboxes[_chain];
-        bytes32 inboxVal = inbox.value;
-
-        for (uint256 i = 0; i < messageCount; i++) {
-            uint256 messageLength = _messageLengths[i];
-            bytes memory messageData = _data.slice(dataOffset, messageLength);
-            dataOffset += messageLength;
-
-            bytes32 messageHash = deliverTransactionSingle(
-                _chain,
-                _tos[i],
-                _seqNumbers[i],
-                _values[i],
-                messageData,
-                _signatures,
-                i * 65
-            );
-
-            inboxVal = Protocol.addMessageToInbox(inboxVal, messageHash);
+        uint256 messageCount = tos.length;
+        require(seqNumbers.length == messageCount, "wrong input length");
+        require(values.length == messageCount, "wrong input length");
+        require(dataLengths.length == messageCount, "wrong input length");
+
+        bytes32 messageHash;
+        assembly {
+            let ptr := mload(0x40)
+            mstore8(ptr, TRANSACTION_BATCH_MSG)
+            ptr := add(ptr, 1)
+            calldatacopy(ptr, 4, sub(calldatasize, 4))
+            ptr := add(ptr, sub(calldatasize, 4))
+            mstore(ptr, number)
+            ptr := add(ptr, 32)
+            mstore(ptr, timestamp)
+            ptr := add(ptr, 32)
+            messageHash := keccak256(mload(0x40), sub(ptr, mload(0x40)))
         }
 
-        inbox.value = inboxVal;
-        inbox.count += messageCount;
-    }
-
-    function deliverTransactionSingle(
-        address _chain,
-        address _to,
-        uint256 _seqNumber,
-        uint256 _value,
-        bytes memory _messageData,
-        bytes memory _signatures,
-        uint256 signatureOffset
-    )
-        private
-        returns(bytes32)
-    {
-        address from = SigUtils.recoverAddress(
-            keccak256(
-                abi.encodePacked(
-                    _chain,
-                    _to,
-                    _seqNumber,
-                    _value,
-                    _messageData
-                )
-            ),
-            _signatures,
-            signatureOffset
-        );
-
-        bytes32 messageHash = Messages.transactionHash(
-            _chain,
-            _to,
-            from,
-            _seqNumber,
-            _value,
-            _messageData,
-            block.number,
-            block.timestamp
-        );
-
+        _deliverMessage(chain, messageHash);
 
-        emit IGlobalInbox.TransactionMessageDelivered(
-            _chain,
-            _to,
-            from,
-            _seqNumber,
-            _value,
-            _messageData
-        );
-
-        return messageHash;
+        emit TransactionMessageBatchDelivered(chain);
```

Comment:
This was done to lower the gas cost of the event. Instead we parse the data directly out of the input to the transaction. However I just realized that this will break if the `deliverTransactionBatch` method is called by a contract rather than an external account since it would lead to the call data of the transaction being arranged differently, breaking our parsing.

I'm not totally sure what the best way to fix this is, but it's a pretty large savings in gas cost so I'd like to maintain the optimization if possible. Technically it's 100% possible to handle, but implementing a universal solution could be difficult